README.md

```diff
@@ -2,4 +2,61 @@
 
 A multi-threaded implementation of a PromQL Query Engine based on the [Volcano/Iterator model](https://paperhub.s3.amazonaws.com/dace52a42c07f7f8348b08dc2b186061.pdf).
 
-The project is currently under development.
+The project is currently under active development.
+
+## Design
+
+At the beggining of a PromQL query execution, the query engine computes a physical plan consisting of multiple independent operators, each responsible for calculating one part of the query expression.
+
+Operators are assembled in a tree-like structure with every operator calling `Next()` on its dependants until there is no more data to be returned. The result of the `Next()` function is a _column vector_  (also called a _step vector_) with elements in the vector representing samples with the same timestamp from different time series.
+
+<p align="center">
+  <img src="./assets/design.png" />
+</p>
+
+This model allows for samples from individual time series to flow one execution step at a time from the left-most operators to the one at the very right. Since most PromQL expressions are aggregations, samples are reduced in number as they are pulled by the operators on the right. Because of this, samples from original timeseries can be decoded and kept in memory in batches instead of being fully expanded.
+
+In addition to operators that have a one-to-one mapping with PromQL constructs, the Volcano model also describes so-called Exchange operators which can be used for flow control and optimizations, such as concurrency or batched selects. An example of an _Exchange_ operator is described in the [Intra-operator parallelism](#intra-operator-parallelism) section.
+
+## Inter-operator parallelism
+
+Since operators are independent and rely on a common interface for pulling data, they can be run in parallel to each other. As soon as one operator has processed data from an evaluation step, it can pass the result onward so that its upstream can immediatelly start working on it.
+
+<p align="center">
+  <img src="./assets/promql-pipeline.png" />
+</p>
+
+
+## Intra-operator parallelism
+
+Parallelism can also be added within individual operators using a parallel coalesce exchange operator. Such exchange operators are indistinguishable from regular operators to their upstreams since they respect the same `Next()` interface.
+
+<p align="center">
+  <img src="./assets/parallel-coalesce.png" />
+</p>
+
+
+## Memory management 
+
+### Step vector allocations
+
+One challenge with the streamed execution model is knowing how much memory to allocate in each operator for each step. 
+
+To work around this issue, operators expose a `Series()` method which returns the labels for all time series that they will ever produce. Operators at the very bottom of the tree, like vector and matrix selectors, have this information since they are responsible for loading data from storage. Other operators can then call `Series()` on the downstream operator and pre-compute all possible outputs.
```

Comment:
```suggestion
To work around this issue, operators expose a `Series()` method which returns the labels for all time series that they will ever produce (for all `Next()` calls). Operators at the very bottom of the tree, like vector and matrix selectors, have this information since they are responsible for loading data from storage. Other operators can then call `Series()` on the downstream operator and pre-compute all possible outputs.
```

Reply:
Am I right?

Reply:
If this is true I am not sure this is useful. With high churn I could have 100 million series in `Series` but `Next` would give me batches at max series at  the current moment which could be smaller e.g. 10k series.

Reply:
I agree that for high churn, this is not that useful. But I would assume most cases are not that high churn so we would have good estimates. However, we can iterate based on real-world data, now we only have synthetic benchmarks.

Reply:
Also worth keeping in mind is that the result always has to have all series. So maybe we can only use this information when allocating vectors for the final result, and not for intermediate operators. Actually each operator might decide whether to use this information or not.

Reply:
> Also worth keeping in mind is that the result always has to have all series.

not really, what if we are grouping?

Reply:
What I meant is that the result always has to have all series that would ever appear in the grouping, regardless of whether there is churn or not.